pkg/quarterdeck/middleware/auth.go

```diff
@@ -68,11 +70,43 @@ func Authenticate(opts ...AuthOption) (_ gin.HandlerFunc, err error) {
 			claims      *tokens.Claims
 		)
 
-		// Get access token from the request
+		// Get access token from the request, if not available then attempt to refresh
+		// using the refresh token cookie.
 		if accessToken, err = GetAccessToken(c); err != nil {
-			log.Debug().Err(err).Msg("no access token in authenticated request")
-			c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
-			return
+			switch {
+			case errors.Is(err, ErrNoAuthorization) && conf.refresher != nil:
+				if cookie, err := c.Cookie(RefreshTokenCookie); err == nil {
+					// Check if the refresh token is still valid
+					if _, err = validator.Verify(cookie); err != nil {
+						log.Debug().Err(err).Msg("could not verify refresh token")
+						c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+						return
+					}
+
+					// Attempt to refresh the access token
+					var refreshToken string
+					if accessToken, refreshToken, err = conf.refresher.Refresh(c.Request.Context(), cookie); err != nil {
+						log.Debug().Err(err).Msg("could not refresh access token")
+						c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+						return
+					}
+
+					// Set the new access and refresh cookies
+					if err = SetAuthCookies(c, accessToken, refreshToken, conf.CookieDomain); err != nil {
+						log.Debug().Err(err).Msg("could not set auth cookies")
+						c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+						return
+					}
+				} else {
+					log.Debug().Err(err).Msg("no access or refresh token in authenticated request")
+					c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+					return
+				}
+			default:
+				log.Debug().Err(err).Msg("no access token in authenticated request")
+				c.AbortWithStatusJSON(http.StatusUnauthorized, api.ErrorResponse(ErrAuthRequired))
+				return
+			}
```

Comment:
I think this can be simplified by a creating a closure function; my next commit will have the example code. If you don't like it, feel free to revert.